File: nokialcd.ts
enum Plots {
    //% block="line"
    Line = 0,
    //% block="box"
    Box = 1,
    //% block="rectangle"
    Rect = 2
}
enum Scrolls {
    //% block="Up"
    Up = 0,
    //% block="Right"
    Right = 1,
    //% block="Down"
    Down = 2,
    //% block="Left"
    Left = 3
}
enum LCDDisplayModes {
    //% block="Normal"
    Normal = 0,
    //% block="Blank"
    Blank = 1,
    //% block="All on"
    AllOn = 2,
    //% block="Inverse"
    Inverse = 3
}

//% weight=100 color=#0fbc11
namespace nokialcd {
    const FILL_X = hex`fffefcf8f0e0c08000`
    const FILL_B = hex`0103070f1f3f7fffff`
    const TWOS = hex`0102040810204080`
    const LCD_CMD = 0
    const LCD_DAT = 1
    const LCD_RST = DigitalPin.P8
    const LCD_CE = DigitalPin.P12
    const LCD_DC = DigitalPin.P16
    const LCD_TEMP = 0x00
    const LCD_BIAS = 0x03
    const LCD_VOP = 0x3f
    let LCD_DE: number = 0x00
    let bytearray: Buffer = initBuffer()

    //% shim=nokialcd::initBuffer
    function initBuffer(): Buffer {
        return pins.createBuffer(504)
    }
    //% shim=nokialcd::SPIinit
    function SPIinit(): void {
        return
    }
    //% blockId=nokialcd_write_byte
    //% block="LCD Wrt  %data"
    export function writeByte(data: number): void {
        pins.digitalWritePin(LCD_CE, 0)
        pins.spiWrite(data)
        pins.digitalWritePin(LCD_CE, 1)
    }
    //% shim=nokialcd::writeFunctionSet
    function writeFunctionSet(v: number, h: number): void {
        pins.digitalWritePin(LCD_DC, LCD_CMD)
        let byte = 0x20 | (v << 1) | (h & 1)
        writeByte(byte)
        pins.digitalWritePin(LCD_DC, LCD_DAT)
    }
    export function init(): void {
        basic.showNumber(1)
        SPIinit()
        basic.showNumber(2)
        writeFunctionSet(0, 1)
        basic.showNumber(3)
        lcdExtendedFunctions(0, 3, 63)
        writeFunctionSet(0, 0)
        lcdDisplayMode(2)
        setX(0)
        setY(0)
        setState(true)
        basic.showNumber(4)
    }

    //% shim=nokialcd::writeSPIBuf
    function writeSPIBuf(): void {
        return
    }

    //% blockId=nokialcd_show
    //% block="show"
    export function show(): void {
        setY(0)
        writeSPIBuf()
    }

    //% shim=TD_ID
    //% blockId="dir_conv" block="%dir"
    //% blockHidden=true
    export function dirs(dir: Scrolls): number {
        return (dir || 0)
    }

    //% shim=TD_ID
    //% blockId="displaymode_conv" block="%displaymode"
    //% blockHidden=true
    export function lcddm(displaymode: LCDDisplayModes): number {
        return (displaymode || 0)
    }

    //% shim=TD_ID
    //% blockId="plot_conv" block="%plot"
    //% blockHidden=true
    export function pl(plot: Plots): number {
        return (plot || 0)
    }

    //% blockId=nokialcd_pixel
    //% block="pixel at x %x y %y %state"
    //% state.shadow="toggleOnOff" state.defl=true
    //% inlineInputMode=inline
    //% shim=nokialcd::pixel
    export function pixel(x: number, y: number, state: boolean): void {
        return
    }

    //% shim=nokialcd::scrollRow
    //% block="scroll row %row direction %direction=dir_conv step %step"
    export function scrollRow(row: number, direction: number, step: number): void {
        return
    }

    //% blockId=nokialcd_display_row
    //% block="show row %row"
    export function displayRow(row: number): void {
    }

    //% blockId=nokialcd_scroll
    //% block="scroll display %direction=dir_conv"
    export function scroll(direction: number): void {
        if (direction & 1) {
            if (direction & 2) {
                // left
                for (let i = 0; i < 503; i++) {
                    bytearray[i] = bytearray[i + 1]
                }
                bytearray[83] = 0
                bytearray[167] = 0
                bytearray[251] = 0
                bytearray[335] = 0
                bytearray[419] = 0
                bytearray[503] = 0
            } else {
                // right
                for (let i = 503; i > 0; i--) {
                    bytearray[i] = bytearray[i - 1]
                }
                bytearray[0] = 0
                bytearray[84] = 0
                bytearray[164] = 0
                bytearray[252] = 0
                bytearray[336] = 0
                bytearray[420] = 0
            }
        } else {
        }
    }

    //% shim=nokialcd::setState
    function setState(s: boolean) {
        return
    }

    //% shim=nokialcd::pLine
    function pLine(x0: number, y0: number, x1: number, y1: number): void {
        return
    }

    //% blockId=nokialcd_plot
    //% block="draw %plot=plot_conv from x %x0 y %y0 to x %x1 y %y1 $state"
    //% state.shadow="toggleOnOff" state.defl=true
    //% inlineInputMode=inline
    export function plot(plot: Plots, x0: number, y0: number, x1: number, y1: number, state: boolean): void {
        setState(state)
        switch (plot) {
            case 0: { pLine(x0, y0, x1, y1); break }
            case 1: { pBox(x0, y0, x1, y1); break }
            case 2: {
                hLine(x0, x1, y0)
                hLine(x0, x1, y1)
                vLine(x0, y0, y1)
                vLine(x1, y0, y1)
                break
            }
            default: pLine(x0, y0, x1, y1)
        }
    }




    //% shim=nokialcd::vLine
    function vLine(x: number, y0: number, y1: number): void {
        return
    }

    //% shim=nokialcd::hLine
    function hLine(x0: number, x1: number, y: number): void {
        return
    }

    //% shim=nokialcd::pBox
    function pBox(x0: number, y0: number, x1: number, y1: number): void {
        return
    }

    //% blockId=nokialcd_clear
    //% block="clear screen"
    export function clear(): void {
        bytearray.fill(0)
        writeBuffer(bytearray)
    }



    //% blockId=nokialcd_write_buffer
    //% block="LCD Wrt %data"
    export function writeBuffer(data: Buffer): void {
        pins.digitalWritePin(LCD_CE, 0)
        for (let i = 0; i < data.length(); i++) {
            pins.spiWrite(data[i])
        }
        pins.digitalWritePin(LCD_CE, 1)
    }

    //% blockId=nokialcd_sety
    //% block="set y address %y"
    //% shim=nokialcd::setYAddr
    export function setY(y: number): void {
        return
    }

    //% blockId=nokialcd_setx
    //% block="set X address %x"
    //% shim=nokialcd::setXAddr
    export function setX(x: number): void {
        return
    }
    //% shim=nokialcd::init2
    export function init2(): void {
        return
    }


    //% blockId=nokialcd_display_mode
    //% block="set display mode %mode"
    export function lcdDisplayMode(mode: number): void {
        pins.digitalWritePin(LCD_DC, LCD_CMD)
        LCD_DE = ((mode & 2) << 1) + (mode & 1)
        writeByte(0x08 | LCD_DE)
        pins.digitalWritePin(LCD_DC, LCD_DAT)

    }

    export function lcdExtendedFunctions(temp: number, bias: number, vop: number): void {
        pins.digitalWritePin(LCD_DC, LCD_CMD)
        writeByte(0x04 | (0x03 & temp))
        writeByte(0x10 | (0x07 & bias))
        writeByte(0x80 | (0x7f & vop))
    }


}
basic.showNumber(1)
nokialcd.init()

File: nokialcd.cpp
#include "pxt.h"
#define mbit_p8                     P0_18       //PIN 18
#define mbit_p12                    P0_20       //PIN 20
#define mbit_p13                    P0_23       //SCK
#define mbit_p14                    P0_22       //MISO
#define mbit_p15                    P0_21       //MOSI
#define mbit_p16                    P0_16       //PIN 16
#define LCD_CMD                         0
#define LCD_DAT                         1
#define LCD_TEMP                        0
#define LCD_BIAS                        3
#define LCD_VOP                         63

using namespace pxt;
    
namespace pins
{
    extern SPI* allocSPI();
    extern int  spiWrite(int value);
    extern void spiFrequency(int frequency);
    extern void spiFormat(int bits, int mode);
}

using namespace pins;


namespace nokialcd {

    DigitalOut LCD_CE(mbit_p12);
    DigitalOut LCD_RST(mbit_p8);
    DigitalOut LCD_DC(mbit_p16);
    static Buffer bytearray = NULL;
    static bool state = true;
    static int lcdDE = 0;

    //%
    void setState(bool s) {
        state = s;
    }

    //%
    void writeSPIByte(int b) {
        auto spi = allocSPI();
        LCD_CE = 0;
        spi->write(b);
        LCD_CE = 1;
    }
    //%
    void writeSPIBuf() {
        auto spi = allocSPI();
        LCD_CE = 0;
        uint8_t *data = bytearray->data;
        for (int i = 0; i < 504; i++, data++) {
            spi->write( *data);
        }
        LCD_CE = 1;
    }
    //%
    void setYAddr(int y) {
        LCD_DC = LCD_CMD;
        writeSPIByte(0x40 + y);
        LCD_DC = LCD_DAT;
    }

    //%
    void setXAddr(int x) {
        LCD_DC = LCD_CMD;
        writeSPIByte(0x80 + x);
        LCD_DC = LCD_DAT;
    }


    void lcdDisplayMode(int mode) {
        lcdDE = ((mode & 2) << 1) + (mode & 1);
        LCD_DC = LCD_CMD;
        writeSPIByte(0x08 | lcdDE);
        LCD_DC = LCD_DAT;
    }
    void writeFunctionSet(int v, int h) {
        auto spi = allocSPI();
        LCD_DC = LCD_CMD;
        LCD_CE = 0;
        spi->write(0x20 | (v << 1) | (h & 1));
        LCD_CE = 1;
        LCD_DC = LCD_DAT;
    }    
    void lcdExtendedFunctions(int temp, int bias, int vop) {
        LCD_DC = LCD_CMD;
        writeSPIByte(0x21);
        writeSPIByte(0x04 | (0x03 & temp));
        writeSPIByte(0x10 | (0x07 & bias));
        writeSPIByte(0x80 | (0x7f & vop));
        writeSPIByte(0x20);
        LCD_DC = LCD_DAT;
    }
    //%
    void SPIinit() {
        LCD_CE = 1;
        lcdDE = 0;
        LCD_RST = 0;
        spiFormat(8,0);
        spiFrequency(1000000);
        wait(0.5);
        LCD_RST = 1;
        writeFunctionSet(0, 1);
        lcdExtendedFunctions(0, 3, 63);
        writeFunctionSet(0, 0);
        lcdDisplayMode(2);
        setXAddr(0);
        setYAddr(0);
        setState(true);
    }
    //%
    void writeBufToLCD() {
        setYAddr(0);
        writeSPIBuf();
    }
    //%
    Buffer initBuffer() {
        bytearray = mkBuffer(NULL,504);
        return bytearray;
    }

    //%
    void pixel(int x, int y, bool state) {
        if (x > 83) {
                return;
                }
        if (y > 47) {
            return;
            }
        if ((x | y) < 0) {
            return;
            }
        uint8_t qy = 1 << (y & 7);
        int r = x + 84 * (y >> 3);
        if (state) {
        bytearray->data[r] |=  qy;
        } else {
        bytearray->data[r] &=  ~qy;
        }
    }

    //%
    void scrollRow(int row, int direction, int step) {
        if ((row < 0) || (row > 5)) return;
        row *= 84;
        int r1 = row + 83;
        if (direction & 1) {
            if (direction & 2) {
                // left
                for (; row < r1; row++) {
                    bytearray->data[row] = bytearray->data[row + 1];
                }
                bytearray->data[row] = 0;
            } else {
                // right
                for (; r1 > row; r1--) {
                    bytearray->data[r1] = bytearray->data[r1 - 1];
                }
                bytearray->data[row] = 0;
            }
        }
    }
    
    //%
    void scrollRowStep(int row, int direction, int step) {
        if ((row < 0) || (row > 5)) return;
        row *= 84;
        int r1 = row + 83;
        if (direction & 1) {
            if (direction & 2) {
                // left
                for (; row < r1; row++) {
                    bytearray->data[row] = bytearray->data[row + 1];
                }
                bytearray->data[row] = 0;
            } else {
                // right
                for (; r1 > row; r1--) {
                    bytearray->data[r1] = bytearray->data[r1 - 1];
                }
                bytearray->data[row] = 0;
            }
        }
    }

    //%
   void vLine(int x, int y0, int y1) {
        int y = y0;
        if (y0 > y1) {
            y = y1;
            y1 = y0;
            }
        if (((x | y1) < 0) || (x > 83)) return;
        if (y < 0) y = 0;
        if (y1 > 47) y1 = 47;
        uint8_t bitmask = 0xff << (y & 7);
        int r = x + 84 *(y >> 3);
        if ((y >> 3) == (y1 >> 3)) {
            bitmask ^= 0xfe <<  (y1 & 7);
            if (state) bytearray->data[r] |= bitmask;
            else bytearray->data[r] &= ~bitmask;
        }
        else {
            int j = (y1 >> 3) - (y >> 3) - 1;
            if (state) bytearray->data[r] |= bitmask;
            else bytearray->data[r] &= ~bitmask;
            r = r + 84;
            if (state) {
                while (j > 0) {
                    bytearray->data[r] = 0xff;
                    j -= 1;
                    r = r + 84;
                }
            } else {
                while (j > 0) {
                    bytearray->data[r] = 0;
                    j -= 1;
                    r = r + 84;
                }
            }
            bitmask = (2 << (y1 & 7)) - 1;
            if (state) bytearray->data[r] |= bitmask;
            else bytearray->data[r] &= ~bitmask;
        }
    }
    
    //%
    void hLine(int x0, int x1, int y) {
        int x = x0;
        if (x0 > x1) { x = x1; x1 = x0; }
        if (((x1 | y) < 0) || (y > 47)) return;
        if (x < 0) x = 0;
        if (x1 > 83) x1 = 83;
        uint8_t bitmask = 1 << (y & 7);
        int r = x + 84 * (y >> 3);
        if (state) {
            for (; x <= x1; x++ , r++)  bytearray->data[r] |= bitmask;
        } else {
            bitmask = ~bitmask;
            for (; x <= x1; x++ , r++) bytearray->data[r] &= bitmask;
        }
    }

    //%
    void pLine(int x0, int y0, int x1, int y1) {
        int dx = abs(x1 - x0);
        if (dx == 0) {
            vLine(x0, y0, y1);
            return;
        }
        int dy = abs(y1 - y0);
        if (dy == 0) {
            hLine(x0, x1, y0);
            return;
        }
        int x = x0;
        int y = y0;
        if (dx > dy) {
            if (x0 > x1) { x = x1; y = y1; x1 = x0; y1 = y0; }
            int yc = (y1 > y) ? 1 : -1;
            int mid = (x + x1) >> 1;
            int a = dy << 1;
            int p = a - dx;
            int b = p - dx;
            pixel(x, y, state);
            while (x < x1) {
                if ((p < 0) || ((p == 0) && (x >= mid))) { p += a; }
                else { p = p + b; y += yc; }
                x++;
                pixel(x, y, state);
            }
        } else {
            if (y0 > y1) { x = x1; y = y1; x1 = x0; y1 = y0; }
            int xc = (x1 > x) ? 1 : -1;
            int mid = (y + y1) >> 1;
            int a = dx << 1;
            int p = a - dy;
            int b = p - dy;
            pixel(x, y, state);
            while (y < y1) {
                if ((p < 0) || ((p == 0) && (y >= mid))) { p += a; }
                else { p = p + b; x += xc; }
                y++; pixel(x, y, state);
            }
        }
    }

    //%
    void pBox(int x0, int y0, int x1, int y1) {
        int x = x0;
        int y = y0;
        if (x1 < x0) { x = x1; x1 = x0; }
        if (y1 < y0) { y = y1; y1 = y0; }
        if ((y1 | x1) < 0) {
            return;
        }
        if (y < 0) y = 0;
        if (y1 > 47) y1 = 47;
        if (x1 > 83) x1 = 83; 
        uint8_t bitmask = 0xff << (y & 7);
        y >>= 3;
        if (y == (y1 >> 3)) {
            bitmask ^= 0xfe << (y1 & 7);
            int r = 84 * y + x;
            if (state) {
                for (; x <= x1; x++ , r++) { bytearray->data[r] |= bitmask; }
            } else {
                bitmask = ~bitmask;
                for (; x <= x1; x++ , r++) { bytearray->data[r] &= bitmask; }
            }
        }
        else {
            int j = (y1 >> 3) - y - 1;
            int r = 84 * y;
            if (state) {
                for (int i = x; i <= x1; i++)  { bytearray->data[i + r] |= bitmask; }
                r += 84;
                while (j > 0) {
                    for (int i = x; i <= x1; i++)  { bytearray->data[i + r] = 0xff; }
                    j--;
                    r += 84;
                }
                bitmask = (2 << (y1 & 7)) - 1;
                for (int i = x; i <= x1; i++) { bytearray->data[i + r] |= bitmask; }
            } else {
                bitmask = ~bitmask;
                for (int i = x; i <= x1; i++) { bytearray->data[i + r] &= bitmask; }
                r += 84;
                while (j > 0) {
                    for (int i = x; i <= x1; i++) { bytearray->data[i + r] = 0; }
                    j--;
                    r += 84;
                }
                bitmask = 0xfe << (y1 & 7);
                for (int i = x; i <= x1; i++) { bytearray->data[i + r] &= bitmask; }
            }
        }
    }
}